src/grpc/qgrpcchannel.cpp

```diff
@@ -0,0 +1,382 @@
+
+        QString rpcName = QString("/%1/%2").arg(service).arg(subscription->method());
+        QGrpcChannelSubscription *sub = new QGrpcChannelSubscription(subscription);
+
+        std::shared_ptr<QMetaObject::Connection> abortConnection(new QMetaObject::Connection);
+        std::shared_ptr<QMetaObject::Connection> readConnection(new QMetaObject::Connection);
+        std::shared_ptr<QMetaObject::Connection> connection(new QMetaObject::Connection);
+
+        grpcWorker.Subscribe(rpcName, subscription->arg(), sub);
+
+        *readConnection = QObject::connect(sub, &QGrpcChannelSubscription::data, subscription, [subscription](const QByteArray &data) {
+            subscription->handler(data);
+        });
+
+        *connection = QObject::connect(sub, &QGrpcChannelSubscription::finished, subscription, [sub, subscription, readConnection, abortConnection, this, service, client](){
+            qProtoDebug() << "Subscription ended with server closing connection";
+
+            if (*readConnection) {
+                QObject::disconnect(*readConnection);
+            }
+            if (*abortConnection) {
+                QObject::disconnect(*abortConnection);
+            }
+
+            switch (sub->status.code()) {
+                case QGrpcStatus::Unavailable:
+                    qProtoDebug() << "Remote server closed connection. Reconnect silently";
+                    subscribe(subscription, service, client);
+                    break;
+                case QGrpcStatus::Ok:
+                    break;
+                default:
+                    subscription->error(sub->status);
+                    break;
+            }
+        });
+
+        *abortConnection = QObject::connect(subscription, &QGrpcSubscription::finished, sub, [connection, abortConnection, readConnection, sub] {
+            qProtoDebug() << "Subscription client was finished";
+
+            if (*connection) {
+                QObject::disconnect(*connection);
+            }
+            if (*readConnection) {
+                QObject::disconnect(*readConnection);
+            }
+            if (*abortConnection) {
+                QObject::disconnect(*abortConnection);
+            }
+
+            sub->terminate();
+        });
+
+        QObject::connect(client, &QAbstractGrpcClient::destroyed, sub, [readConnection, connection, abortConnection, sub](){
+            qProtoDebug() << "Grpc client was destroyed";
+
+            if (*readConnection) {
+                QObject::disconnect(*readConnection);
+            }
+            if (*connection) {
+                QObject::disconnect(*connection);
+            }
+            if (*abortConnection) {
+                QObject::disconnect(*abortConnection);
+            }
+
+            sub->terminate();
+        });
+
+        sub->start();
+    }
+};
+
+}
+
+QGrpcChannel::QGrpcChannel(const QUrl &url, std::unique_ptr<QAbstractGrpcCredentials> credentials) : QAbstractGrpcChannel()
```

Comment:
For credentials please use native grpc credentials here directly instead. As per design constructor might be called by used with any parameters, so it's better to use grpc::ChannelCredentials here